### PR TITLE
Extract packages in background

### LINKED ISSUE #29: Extract while other packages are downloading
When needing to fetch and extract `n` packages, extract `n-1` while downloading `n`, and so forth.

### PR DESCRIPTION
**Pull request by tbateson**
Closes #29

### CODE REVIEW

File: src/pwr.ps1 [hunk 1]
```diff
@@ -429,21 +432,29 @@ function Test-PwrPackage($Pkg) {
...
 		}
-		Write-PwrHost 'done.'
+	}
+	return $Jobs
+}
+
+function Receive-PwrJobs($Jobs) {
+	if ($Jobs.Count -gt 0) {
+		Receive-Job -Job $Jobs -Wait -AutoRemoveJob
+		Write-PwrOutput "fetched $($Packages.Count) package$(if ($Packages.Count -ne 1) { 's' })"
```
[the remark this answers is not in the document]
Comment (tbateson) [reply src/pwr.ps1:457]: Yeah I unfortunately realized this, working on a fix - not as easy as I would like it to be.

File: src/pwr.ps1 [hunk 2]
```diff
@@ -833,6 +846,7 @@ switch ($Command) {
 					Write-PwrOutput "$($Pkg.Ref) already loaded"
 				}
 			}
+			Receive-PwrJobs $Jobs
 			$env:Path = "$env:Path;$_Path"
 		} catch {
 			$env:Path = $_Path
```
[the remark this answers is not in the document]
Comment (tbateson) [reply src/pwr.ps1:852]: Shouldn't we `exit 1` or `Write-PwrFatal`? Nothing to catch if we throw here.

File: src/pwr.ps1 [hunk 3]
```diff
@@ -714,7 +727,7 @@ function Invoke-PwrScripts {
 				}
 			}
 			if ($ErrorMessage) {
-				Write-PwrFatal "script '$Name' failed to execute$(if ($ErrorMessage.Length -gt 0) { '' } else { ', ' + $ErrorMessage.Substring(0, 1).ToLower() + $ErrorMessage.Substring(1) })"
+				Write-PwrFatal "script '$Name' failed to execute$(if ($ErrorMessage.Length -eq 0) { ', ' + $ErrorMessage.Substring(0, 1).ToLower() + $ErrorMessage.Substring(1) })"
```
[the remark this answers is not in the document]
Comment (tbateson) [reply src/pwr.ps1:729]: I guess it's been wrong the whole time.

File: src/pwr.ps1 (hunk 2 above)
[the remark this answers is not in the document]
Comment (tbateson) [reply src/pwr.ps1]: Should be fixed

File: src/pwr.ps1 (hunk 1 above)
[the remark this answers is not in the document]
Comment (tbateson) [reply src/pwr.ps1]: Should be reflective of the right number of packages fetched

File: src/pwr.ps1 (hunk 3 above)
[the remark this answers is not in the document]
Comment (tbateson) [reply src/pwr.ps1]: Fixed

File: src/pwr.ps1 (hunk 2 above)
[the remark this answers is not in the document]
Comment (tbateson) [reply src/pwr.ps1]: Currently is `exit 1` in `main`, left it as is.

**Review by troyaf — APPROVED** [review]

File: src/pwr.ps1 [hunk 4]
```diff
@@ -252,10 +252,17 @@ function Invoke-PullImageLayer($Out, $Repo, $Digest) {
...
 		Invoke-PwrWebRequest "$($Repo.Uri)/blobs/$Digest" -OutFile $Tmp -Headers $Headers
 	} else {
-		Write-PwrHost 'using cache ... ' -NoNewline
+		Write-PwrDebug "using cache for $Ref"
 	}
-	& "$env:HOMEDRIVE\Windows\System32\tar.exe" -xzf $Tmp -C $Out --exclude 'Hives/*' --strip-components 1
-	Remove-Item $Tmp
+	$Hash = "sha256:$((Get-FileHash -Path $Tmp -Algorithm SHA256).Hash)"
+	if ($Hash -ne $Digest) {
+		Write-PwrFatal "layer of $Ref with digest $Hash does not match expected digest $Digest`n     the file has been modified, implying a man-in-the-middle or similar attack"
```
Comment (troyaf) [inline src/pwr.ps1:259]: This wording is a bit alarmist. Can we just say something like `the download was corrupted` instead?

Comment (tbateson) [reply src/pwr.ps1]: The wording should be alarmist, but I can see why we might not want it to be. I've changed it.

**Pull request merged**